File: src/main/java/io/github/bloodnighttw/WhateverBot/utils/command/CommandRegister.java
package io.github.bloodnighttw.WhateverBot.utils.command;

import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;

public class CommandRegister {

  private JDA bot;
  private final Logger logger = LoggerFactory.getLogger(CommandRegister.class);
  private ArrayList<CommandData> commandDataList;

  public CommandRegister(JDA bot) {
    this.bot = bot;
    commandDataList = new ArrayList<>();
  }

  public void registerCommand(ICommand iCommand) {
    if (iCommand.isGlobal()) {
      commandDataList.add(iCommand.getCommand());
      for (String alias : iCommand.getAlias()) {
        commandDataList.add(new CommandData(alias, iCommand.getCommand().getDescription()));
      }
    }

    bot.addEventListener(new CommandHandler(iCommand));

    logger.debug("Add Command....");
  }

  public void addToAllServer() {
    for (Guild guild : bot.getGuilds()) {
      logger.debug("Add command to server: " + guild.getName());
      guild.updateCommands().addCommands(commandDataList).queue();
    }
  }
}

File: src/main/java/io/github/bloodnighttw/whateverBot/utils/command/CommandRegister.kt
package io.github.bloodnighttw.whateverBot.utils.command


import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.interactions.commands.build.CommandData
import org.slf4j.LoggerFactory

class CommandRegister(private val bot: JDA) {
	private val logger = LoggerFactory.getLogger(CommandRegister::class.java)
	private val commandDataLocalList: ArrayList<CommandData> = ArrayList()
	private val commandDataGlobalList: ArrayList<CommandData> = ArrayList()

	private fun register(iCommand: ICommand) {
		var tempCommandData = iCommand.command
		val tempDataList: ArrayList<CommandData> =
			if (iCommand.isGlobal) commandDataGlobalList else commandDataLocalList

		val allSubCommand = iCommand.subCommandMap.values.map { it.subCommand }
		tempCommandData = tempCommandData.addSubcommands(allSubCommand)

		tempDataList.add(tempCommandData)

		for (alias in iCommand.alias) {
			tempCommandData =
				CommandData(alias, tempCommandData.description).addSubcommands(allSubCommand)
			tempDataList.add(tempCommandData)
		}

	}

	fun registerCommand(iCommand: ICommand) {
		register(iCommand)
		bot.addEventListener(CommandHandler(iCommand))
	}

	fun registerPrivateCommand(iCommand: ICommand) {
		register(iCommand)
		bot.addEventListener(CommandHandler(iCommand, true))
	}

	fun addToAllServer() {
		bot.updateCommands().addCommands(commandDataGlobalList).queue()
		commandDataLocalList.forEach { logger.debug("Local Command " + it.name) }
		commandDataGlobalList.forEach { logger.debug("Global Command " + it.name) }
		for (guild in bot.guilds) {
			logger.debug("Guild Found Name:" + guild.name)
			logger.debug("Add command to server: " + guild.name)
			guild.updateCommands()
				.addCommands(commandDataLocalList)
				.addCommands(commandDataGlobalList)
				.queue()
		}
	}

}